src/webview/css/css.js

```diff
@@ -1,11 +1,76 @@
 /* @flow strict-local */
+import { Platform } from 'react-native';
 import type { ThemeName } from '../../types';
 import cssPygments from './cssPygments';
 import cssEmojis from './cssEmojis';
 import cssNight from './cssNight';
 
+/** CSS fragment to support scrolling of KaTeX formulae. */
+/*
+  By default, KaTeX renders (non-inline) math into a div of fixed precomputed
+  width. This will be cut off by the edge of the screen when the formula is too
+  long -- and on a mobile device, that's very nearly always.
+
+  The naïve solution of simply giving some part of the KaTeX fragment itself an
+  `overflow-x: auto` style breaks terribly:
+    * Margin collapsing no longer works, causing rendering artifacts. (This is
+      particularly visible on integral signs, which are truncated into near-
+      illegibility.)
```

Comment:
(Continuing from https://github.com/zulip/zulip-mobile/pull/3679#discussion_r344937078 .)

I appreciate the extra few words here, "which are truncated into near-illegibility".

I still don't understand what is meant by "margin collapsing no longer works", though, and I don't understand what goes wrong if you do this "naive" solution.

One reason I appreciate the extra words is in fact that they provide something concrete, so that I can tell if I'm seeing the same thing you're describing! I did the following experiment:
* Checked out this branch just before this commit -- so there's KaTeX, but not this scrolling fixup.
* Opened the app in the Android emulator, and went to a message list with a display-math integral in it.
* The integral looked fine.
* Opened the Chrome devtools on the webview, and edited the stylesheet so that the `.katex-display` element would have `overflow-x: auto`.
* Result: the integral still looked fine and normal. I didn't see anything truncated.
* Another display formula, this one horizontally long so it overflowed the screen, also looked fine, and scrolling it worked as expected.

As a variation, I made one of the integral endpoints unreasonably tall, in case that was the situation you had particularly in mind. Still showed up fine:
![Screenshot from 2020-03-26 20-01-42](https://user-images.githubusercontent.com/28173/77719781-613d9680-6fa3-11ea-8bd6-4aa04245ad82.png)
The ℵ (aleph) is potentially hard to read because it's small, but that seems expected given where it is in the formula; and in particular nothing appears to be truncated.

Reply:
> I still don't understand what is meant by "margin collapsing no longer works", though

Describing this as being caused by margin collapsing [[1]](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Box_Model/Mastering_margin_collapsing) no longer working [[2]](https://stackoverflow.com/questions/19718634/how-to-disable-margin-collapsing/19719427#19719427), is, I *vaguely* recall, based on a comment in a KaTeX GitHub issue. (Which, naturally, I can no longer find. :disappointed:)

> * Checked out this branch just before this commit -- so there's KaTeX, but not this scrolling fixup.
> * Opened the app in the Android emulator, and went to a message list with a display-math integral in it.
> * The integral looked fine.
> * Opened the Chrome devtools on the webview, and edited the stylesheet so that the `.katex-display` element would have `overflow-x: auto`.

I followed this procedure, and got this:
![Screenshot_1585442774_scaled](https://user-images.githubusercontent.com/54861490/77837343-e6988680-711c-11ea-9412-6392db1cf604.png)
Which is about what I remember seeing. (This would probably not be a bad image to provide a link to (a thousand words, etc.), if GitHub is a reliable-enough imagehost.)

You're probably not seeing it because you're on the latest WebView, where I couldn't reproduce it in this way. (I customarily test on API 21, for essentially this reason.) I think it also breaks on Safari, or at least that it did three months ago; but I'm not sure, and can't presently test it.

Reply:
Ah! Very interesting, thanks -- both the screenshot (a thousand words, etc.), and the repro detail about what platform you tried it on.

I'll go try this on some older versions. I think it's also possible to describe the situation in the comment, and then link to this screenshot as a helpful bonus -- something like: "Some formulas have rendering errors on older browsers (before Chrome NN and Safari MM); in particular, `\int` signs get clipped halfway through."

Another key aspect of this behavior is that it's apparently a browser bug. From the earlier descriptions, I couldn't tell if the issue (as you understood it) was that doing this triggered browser bugs, or that it just interacted badly, through the legitimately-specced arcanities of CSS and of margin-collapsing in particular, with what KaTeX was doing with its CSS.

When the behavior is faithful to the spec, I tend to feel we should understand what's going on better before doing a hack to work around it, because it's likely there's a cleaner way which the spec authors (or people giving them feedback) had in mind. But when it's a browser bug -- and when some broken-seeming behavior is fixed in newer browser versions, that's prima facie evidence it's a browser bug -- then a hack is much more likely to be the best solution available.

Reply:
> API 21

Looking this up, I think you're referring to Android 5.0 Lollipop.

Here's what we have in the "supported platforms" comment in js.js:
```
/*
 * Supported platforms:
 *
 * * We support iOS 10.  So this code needs to work on Mobile Safari 10.
 *   Graceful degradation is acceptable below iOS 12 / Mobile Safari 12.
 *
 * * For Android, core functionality needs to work on Chrome 44.
 *   Graceful degradation is acceptable below Chrome 58.
 *
 *   * These versions are found in stock images for Android 6 Marshmallow
 *     and Android 8 Oreo, respectively, for convenient testing.
 *
 *   * (Note that Android's Chrome auto-updates independently of the OS, and
 *     the large majority of Android users have a fully-updated Chrome --
 *     more recent than the Safari on most iOS devices.)
 *
 *   * Below Chrome 44, it's possible (but rare) for a user to be on a
 *     version as old as Chrome 37, which shipped with Android 5 Lollipop.
 *     We sometimes fix issues affecting those versions, only when trivial.
 *
 * * See docs/architecture/platform-versions.md for data and discussion
 *   about our version-support strategy.
 */
```

So the Chrome 37 that shipped with that release (and that appears in stock emulator images for it) is in the status of "we sometimes fix issues affecting those versions, only when trivial".

I think it'd be quite reasonable to test routinely on Chrome 44, which you can get in an Android 6.0 Marshmallow (aka API 23) stock emulator image. If you test on Chrome 37 and find something broken, it'd be wise to test on Chrome 44 before spending more than a few minutes seeking to fix it.

This particular misrendering, I wouldn't describe as "core functionality" failing to work, either -- it looks ugly, clearly something's broken, but you can still pretty much read everything. So by our policy, it's OK to have (and doesn't justify a complex hack to fix) if it appears only below Chrome 58 (found on an Android 8.0 Oreo stock image) and iOS 12.

Reply:
> Another key aspect of this behavior is that it's apparently a browser bug. From the earlier descriptions, I couldn't tell if the issue (as you understood it) was that doing this triggered browser bugs, or that it just interacted badly, through the legitimately-specced arcanities of CSS and of margin-collapsing in particular, with what KaTeX was doing with its CSS.

Let me clarify then: at the time, I understood it to be the *latter*, and I am not certain even now that it's not. (The fact that Chrome's behavior has changed is certainly evidence in favor, but – CSS being, as you note, CSS – not proof.)

Even if it's not, though – as long as Safari v${latest - 1} (which, as I mentioned, I can't test right now) and stock Android 8.0 (which I can and will) are happy with `overflow: auto`, I expect I will be, too.

My initial tests have [something else going wrong](https://user-images.githubusercontent.com/54861490/77966390-c8a36100-7297-11ea-82cf-3b9be62c0212.png), though, so there may be a bit of a delay there. (This is on Android 10, on a pristine commit 0b3d3b90b362572f3406f8052cce48067c4f96ba, which I'm pretty sure worked on Saturday. :disappointed:)

> So the Chrome 37 that shipped with that release [...] is in the status of "we sometimes fix issues affecting those versions, only when trivial".

Hmm. I'd thought our support for Lollipop was more firm than that.

That said, I typically *do* test interesting bugs against one or more of a small collection of Android images; I just didn't have them to hand, for obvious reasons. If this breaks on Chrome 37 but not Chrome 44, it'll be the first thing I can recall which does.

Reply:
... and, indeed, it *does* break on stock Marshmallow (API 23, Chrome 44). It doesn't break on Oreo (API 27, Chrome 61).

I haven't tested any intermediate images, because it's moot: even on Oreo, the second issue in that comment – the fact that the equations also become vertically scrollable – is still a problem. (I haven't tested an Android 10 image specifically, but it happens on desktop Chrome 80.)

Reply:
Cool. Certainly worth studying more, then. I didn't get to testing this issue today as much as I'd hoped, but I did just now try this:

> even on Oreo, the second issue in that comment – the fact that the equations also become vertically scrollable – is still a problem.

and I don't reproduce. Same procedure as above, this time with a `\sum... \int... \frac...` that looks like one of your examples in the screenshot above, and (on an Android 8 Oreo emulator image) it looks normal and scrolls horizontally but not vertically, as desired.

So, definitely interested in pinning that down.

... Hm! But `navigator.userAgent` in the dev console there tells me it's Chrome 72. Must have auto-updated :-/ -- normally a good thing, but not for this kind of testing. I'll have to revert that, or if nothing else then make a fresh image.

---

Any comments on the branch before this point? I think even without horizontal scrolling of display math, this is a major improvement over the status quo, both for display math and (what I think is more common) inline math. So I would be very pleased to merge those n-2 commits soon, and then have a followup task to get horizontal scrolling for wide display formulas.

Reply:
> and I don't reproduce. Same procedure as above, this time with a `\sum... \int... \frac...` that looks like one of your examples in the screenshot above, and (on an Android 8 Oreo emulator image) it looks normal and scrolls horizontally but not vertically, as desired.

The exact text was:
````
```math
\lim_{x \to \infty} \int_{0}^{1} \frac{1 + 2x + 3x^2 + 4x^3 + 5x^4 + 6x^5 \cdots}{e^x} \; dx
```
````

Here's a shot of it with the equation scrolled upwards slightly. Note the truncated exponents.
![Screenshot_20200330_180355](https://user-images.githubusercontent.com/54861490/77976103-db765f80-72b0-11ea-8ad6-eaa85dc584a2.png)

(The problem is not so much that the exponents are truncated; it's that the equations are all wobbly to the touch. It's not a good look.)

> Any comments on the branch before this point? 

A few things. I had to cancel my review to respond here, but I have them saved; I'll finish them up and post them shortly.

> I think even without horizontal scrolling of display math, this is a major improvement over the status quo, both for display math and (what I think is more common) inline math. So I would be very pleased to merge those n-2 commits soon, and then have a followup task to get horizontal scrolling for wide display formulas.

Wait, what? Why *not* the last two? The existing horizontal-scrolling code is functional everywhere without issues, as far as I can tell. If investigation shows that something simpler is feasible, we can swap it out later without affecting users.

Besides which, I don't think leaving them out leaves the app in a good state. It's possible for an embedded equation to be cut off at a point where it's not obvious that it's not all there.

Reply:
Great -- thanks for the repro, and for the comments on the rest!

> Besides which, I don't think leaving them out leaves the app in a good state. It's possible for an embedded equation to be cut off at a point where it's not obvious that it's not all there.

The webapp has the same behavior. Here's [two messages](https://chat.zulip.org/#narrow/stream/7-test-here/topic/latex/near/844938) with the same content `1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 + 9 + 10 + 11 + 12 + 13 + 14 + 15 + 16 + \cdots`, except one is in an inline math formula and the other in display math:

![image](https://user-images.githubusercontent.com/28173/78073076-0fa35c00-7355-11ea-96b6-51c97cda1176.png)

We've heard very clearly from users that the current state in the mobile app is that LaTeX doesn't work, that they're very happy they can use LaTeX in the webapp, and that they want it in the mobile app too. It'll be good to fix this edge case, but I am not going to block on it for shipping the feature.

Reply:
(Meant to reply directly to this part too, and forgot.)

> Why *not* the last two? The existing horizontal-scrolling code is functional everywhere without issues, as far as I can tell.

Complexity. The app has had a number of complex hacks in it that were made without clearly understanding, or simply without clearly explaining, what problem they were solving and how and why. And it's suffered from that.

There's a place for complex hacks, and there may well be a place for this one. But we don't have a clear enough understanding at this point to be ready to make it.

Reply:
Filed zulip/zulip#14422 for the issue in the existing webapp.